## Feature Engineering 6

Q1. Pearson correlation coefficient is a measure of the linear relationship between two variables. Suppose
you have collected data on the amount of time students spend studying for an exam and their final exam
scores. Calculate the Pearson correlation coefficient between these two variables and interpret the result.

Ans:  
  
### Pearson Correlation Coefficient Calculation Using Covariance

The Pearson correlation coefficient (\( r \)) measures the linear relationship between two variables. It can be expressed using covariance and standard deviations as follows:

$$
r = \frac{\text{Cov}(X, Y)}{\sigma_X \sigma_Y}
$$

where:
- $\text{Cov}(X, Y)$ is the covariance of X and Y,
- $\sigma_X is$ the standard deviation of X,
- $\sigma_Y$ is the standard deviation of \(Y\).

### Example Data

Let's use the following data for our calculation:

| Student | Time Studied (X) | Exam Score (Y) |
|---------|------------------|----------------|
| 1       | 2                | 50             |
| 2       | 3                | 60             |
| 3       | 5                | 70             |
| 4       | 7                | 80             |
| 5       | 8                | 85             |

### Step-by-Step Calculation

#### 1. Calculate the Means

$$
\bar{X} = \frac{2 + 3 + 5 + 7 + 8}{5} = \frac{25}{5} = 5
$$

$$
\bar{Y} = \frac{50 + 60 + 70 + 80 + 85}{5} = \frac{345}{5} = 69
$$

#### 2. Calculate the Covariance

$$
\text{Cov}(X, Y) = \frac{(2 - 5)(50 - 69) + (3 - 5)(60 - 69) + (5 - 5)(70 - 69) + (7 - 5)(80 - 69) + (8 - 5)(85 - 69)}{5}
$$

$$
= \frac{(-3 \cdot -19) + (-2 \cdot -9) + (0 \cdot 1) + (2 \cdot 11) + (3 \cdot 16)}{5}
$$

$$
= \frac{57 + 18 + 0 + 22 + 48}{5}
$$

$$
= \frac{145}{5} = 29
$$

#### 3. Calculate the Standard Deviations

For \(X\):

$$
\sigma_X = \sqrt{\frac{(2 - 5)^2 + (3 - 5)^2 + (5 - 5)^2 + (7 - 5)^2 + (8 - 5)^2}{5}}
$$

$$
= \sqrt{\frac{9 + 4 + 0 + 4 + 9}{5}}
$$

$$
= \sqrt{\frac{26}{5}} \approx \sqrt{5.2} \approx 2.28
$$

For \(Y\):

$$
\sigma_Y = \sqrt{\frac{(50 - 69)^2 + (60 - 69)^2 + (70 - 69)^2 + (80 - 69)^2 + (85 - 69)^2}{5}}
$$

$$
= \sqrt{\frac{361 + 81 + 1 + 121 + 256}{5}}
$$

$$
= \sqrt{\frac{820}{5}} \approx \sqrt{164} \approx 12.81
$$

#### 4. Calculate \( r \)

$$
r = \frac{\text{Cov}(X, Y)}{\sigma_X \sigma_Y} = \frac{29}{2.28 \cdot 12.81}
$$

$$
= \frac{29}{29.16} \approx 0.994
$$

### Summary

The Pearson correlation coefficient \( r \) is approximately \( 0.994 \), which indicates a very strong positive linear relationship between the amount of time students spend studying and their final exam scores. This result suggests that as study time increases, exam scores tend to increase significantly.


Q2. Spearman's rank correlation is a measure of the monotonic relationship between two variables.
Suppose you have collected data on the amount of sleep individuals get each night and their overall job
satisfaction level on a scale of 1 to 10. Calculate the Spearman's rank correlation between these two
variables and interpret the result.

Ans: 
  
#### Spearman's Rank Correlation Coefficient Calculation

Spearman's rank correlation coefficient (\( \rho \)) measures the monotonic relationship between two variables. It is calculated using:

$$
\rho = 1 - \frac{6 \sum d_i^2}{n(n^2 - 1)}
$$

where:
- \( d_i \) is the difference between the ranks of each pair of values,
- \( n \) is the number of data points.

**Example Data**  

Let's use the following revised data on sleep and job satisfaction:

| Individual | Sleep (hours) | Job Satisfaction (1-10) |
|------------|---------------|-------------------------|
| 1          | 5             | 6                       |
| 2          | 6             | 7                       |
| 3          | 8             | 9                       |
| 4          | 7             | 6                       |
| 5          | 4             | 8                       |

**1. Rank the Data**

**Ranks for Sleep:**

| Individual | Sleep (hours) | Rank |
|------------|---------------|------|
| 1          | 5             | 2    |
| 2          | 6             | 3    |
| 3          | 8             | 5    |
| 4          | 7             | 4    |
| 5          | 4             | 1    |

**Ranks for Job Satisfaction:**

| Individual | Job Satisfaction (1-10) | Rank |
|------------|-------------------------|------|
| 1          | 6                       | 2    |
| 2          | 7                       | 3    |
| 3          | 9                       | 5    |
| 4          | 6                       | 2    |
| 5          | 8                       | 4    |

**2. Calculate the Difference in Ranks**

Compute \( d_i \) and \( d_i^2 \):

| Individual | Rank (Sleep) | Rank (Satisfaction) | \( d_i \) | \( d_i^2 \) |
|------------|--------------|---------------------|-----------|-------------|
| 1          | 2            | 2                   | 0         | 0           |
| 2          | 3            | 3                   | 0         | 0           |
| 3          | 5            | 5                   | 0         | 0           |
| 4          | 4            | 2                   | 2         | 4           |
| 5          | 1            | 4                   | -3        | 9           |

**3. Sum of Squared Differences**

$$
\sum d_i^2 = 0 + 0 + 0 + 4 + 9 = 13
$$

**4. Calculate $\rho$**

Substitute into the formula:

$$
\rho = 1 - \frac{6 \sum d_i^2}{n(n^2 - 1)}
$$

where \( n = 5 \).

$$
\rho = 1 - \frac{6 \times 13}{5 \times (5^2 - 1)}
$$

$$
\rho = 1 - \frac{78}{5 \times 24}
$$

$$
\rho = 1 - \frac{78}{120}
$$

$$
\rho = 1 - 0.65
$$

$$
\rho = 0.35
$$

**Interpretation:**  

A Spearman's rank correlation coefficient $\rho$ of 0.35 indicates a moderate positive monotonic relationship between the amount of sleep and job satisfaction. As sleep increases, job satisfaction tends to increase, but the relationship is not as strong as in a perfect monotonic relationship.


Q3. Suppose you are conducting a study to examine the relationship between the number of hours of
exercise per week and body mass index (BMI) in a sample of adults. You collected data on both variables
for 50 participants. Calculate the Pearson correlation coefficient and the Spearman's rank correlation
between these two variables and compare the results.

Ans:  

To calculate both the Pearson correlation coefficient and Spearman's rank correlation coefficient using Python, you'll need to use the pandas library for handling the data and scipy for statistical computations

In [2]:
import numpy as np
import pandas as pd
from scipy.stats import pearsonr, spearmanr

# Generate synthetic data for demonstration
np.random.seed(0)
exercise_hours = np.random.normal(loc=3, scale=1, size=50)  # Mean = 3 hours, SD = 1 hour
bmi = np.random.normal(loc=25, scale=5, size=50)  # Mean = 25 BMI, SD = 5 BMI

# Create a DataFrame
data = pd.DataFrame({
    'Exercise Hours': exercise_hours,
    'BMI': bmi
})

# Calculate Pearson Correlation
pearson_corr, _ = pearsonr(data['Exercise Hours'], data['BMI'])

# Calculate Spearman's Rank Correlation
spearman_corr, _ = spearmanr(data['Exercise Hours'], data['BMI'])

# Print results
print(f"Pearson Correlation Coefficient: {pearson_corr:.4f}")
print(f"Spearman's Rank Correlation Coefficient: {spearman_corr:.4f}")


Pearson Correlation Coefficient: -0.0592
Spearman's Rank Correlation Coefficient: -0.0742


Q4. A researcher is interested in examining the relationship between the number of hours individuals
spend watching television per day and their level of physical activity. The researcher collected data on
both variables from a sample of 50 participants. Calculate the Pearson correlation coefficient between
these two variables.

Ans:  

To calculate the Pearson correlation coefficient between the number of hours individuals spend watching television per day and their level of physical activity, you'll need to follow these steps in Python:

In [11]:
import numpy as np
import pandas as pd
from scipy.stats import pearsonr

# Set a random seed for reproducibility
np.random.seed(0)

# Generate synthetic data with a stronger relationship
# Here we create a linear relationship with added noise
n = 50  # Number of participants
tv_hours = np.linspace(1, 5, n)  # TV hours ranging from 1 to 5
# Generate physical activity that is inversely related to TV hours with some noise
physical_activity = 10 - 2 * tv_hours + np.random.normal(0, 1, n)  # Linear relationship with noise

# Create a DataFrame
data = pd.DataFrame({
    'TV Hours': tv_hours,
    'Physical Activity (hours)': physical_activity
})

# Calculate Pearson Correlation
pearson_corr, p_value = pearsonr(data['TV Hours'], data['Physical Activity (hours)'])

# Print results
print(f"Pearson Correlation Coefficient: {pearson_corr:.4f}")
print(f"P-value: {p_value}")


Pearson Correlation Coefficient: -0.9356
P-value: 2.414147589032731e-23


In [12]:
print(f"P-value: {p_value: .4f}")

P-value:  0.0000


**Interpretation:**  
Pearson Correlation Coefficient: A value of -0.9356 indicates a strong negative linear relationship between TV hours and physical activity. As TV hours increase, physical activity decreases.  
P-value: A p-value of $2.4141*10^{-10}$ is much smaller than 0.05 as it is almost equal to 0, indicating that the correlation is statistically significant. This suggests that the observed relationship is unlikely due to random chance and is a meaningful finding.

Q5. A survey was conducted to examine the relationship between age and preference for a particular
brand of soft drink. The survey results are shown below:

Age = [25,42,37,19,31,28]  
soft drink preference = [coke, pepsi, mountain dew, coke, pepsi, coke]

In [13]:
import numpy as np
import pandas as pd
from scipy.stats import pearsonr
from sklearn.preprocessing import LabelEncoder

# Given data
age = [25, 42, 37, 19, 31, 28]
soft_drink_preference = ['coke', 'pepsi', 'mountain dew', 'coke', 'pepsi', 'coke']

# Encode categorical data
label_encoder = LabelEncoder()
encoded_preferences = label_encoder.fit_transform(soft_drink_preference)

# Create a DataFrame
data = pd.DataFrame({
    'Age': age,
    'Soft Drink Preference': encoded_preferences
})

# Calculate Pearson Correlation
pearson_corr, p_value = pearsonr(data['Age'], data['Soft Drink Preference'])

# Print results with formatting to 2 decimal places
print(f"Pearson Correlation Coefficient: {pearson_corr:.2f}")
print(f"P-value: {p_value:.2f}")


Pearson Correlation Coefficient: 0.77
P-value: 0.07


**Interpretation:**:    
1. Pearson Correlation Coefficient: A value of 0.77 suggests a moderate positive linear relationship between age and the encoded soft drink preferences.
2. P-value: A p-value of 0.07 indicates that the correlation is not statistically significant at the 0.05 level, implying that the observed relationship may not be meaningful or could be due to random chance.

Q6. A company is interested in examining the relationship between the number of sales calls made per day
and the number of sales made per week. The company collected data on both variables from a sample of
30 sales representatives. Calculate the Pearson correlation coefficient between these two variables.

Ans:  

To calculate the Pearson correlation coefficient between the number of sales calls made per day and the number of sales made per week, you'll need to follow these steps:

In [14]:
import numpy as np
from scipy.stats import pearsonr

# Set a random seed for reproducibility
np.random.seed(0)

# Generate synthetic data
n = 30  # Number of sales representatives

# Generate the number of sales calls made per day
# Assume a normal distribution for sales calls
mean_calls = 70
std_calls = 15
sales_calls_per_day = np.random.normal(mean_calls, std_calls, n)

# Generate the number of sales made per week
# Assume a linear relationship with some added noise
mean_sales = 30
std_sales = 5
sales_per_week = 0.5 * sales_calls_per_day + np.random.normal(mean_sales, std_sales, n)

# Calculate Pearson Correlation Coefficient
pearson_corr, p_value = pearsonr(sales_calls_per_day, sales_per_week)

# Print results with formatting to 2 decimal places
print(f"Pearson Correlation Coefficient: {pearson_corr:.2f}")
print(f"P-value: {p_value:.2f}")


Pearson Correlation Coefficient: 0.85
P-value: 0.00


**Interpretation:**  
1. Pearson Correlation Coefficient: A value of 0.85 indicates a strong positive linear relationship between the number of sales calls made per day and the number of sales made per week. This suggests that as the number of sales calls increases, the number of sales made per week tends to increase as well.   
2. P-value: A p-value of 0.00 indicates that the correlation is statistically significant, meaning the observed relationship is very unlikely to be due to random chance.  